Microsoft.Azure.Cosmos/src/Batch/BatchExecutor.cs

```
@@ -0,0 +1,142 @@
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents;
+
+    internal sealed class BatchExecutor
+    {
+        private readonly ContainerCore container;
+
+        private readonly CosmosClientContext clientContext;
+
+        private readonly IReadOnlyList<ItemBatchOperation> inputOperations;
+
+        private readonly PartitionKey partitionKey;
+
+        private readonly RequestOptions batchOptions;
+
+        private readonly int maxServerRequestBodyLength;
+
+        private readonly int maxServerRequestOperationCount;
+
+        public BatchExecutor(
+            ContainerCore container,
+            PartitionKey partitionKey,
+            IReadOnlyList<ItemBatchOperation> operations,
+            RequestOptions batchOptions,
+            int maxServerRequestBodyLength,
+            int maxServerRequestOperationCount)
+        {
+            this.container = container;
+            this.clientContext = this.container.ClientContext;
+            this.inputOperations = operations;
+            this.partitionKey = partitionKey;
+            this.batchOptions = batchOptions;
+            this.maxServerRequestBodyLength = maxServerRequestBodyLength;
+            this.maxServerRequestOperationCount = maxServerRequestOperationCount;
+        }
+
+        public async Task<BatchResponse> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            ResponseMessage validationResult = BatchExecUtils.Validate(
+                this.inputOperations,
+                this.batchOptions,
+                this.maxServerRequestOperationCount);
+
+            if (!validationResult.IsSuccessStatusCode)
+            {
+                return new BatchResponse(

```

Argument exceptions are modeled as Server response. 
Convention is to have ArgumentExceptions for arguments validation

---

Changed this. I am assuming convention is to throw exception if the user input was invalid and so we were not able to make a server call - given this, even if the size of a particular document to be written or overall request size exceeds bounds, we should throw an exception, and I have updated to the same.